code/attention-is-all-you-need/attention.py

```diff
@@ -0,0 +1,65 @@
+import numpy as np
+import math
+
+
+def softmax(x):
+    x = x - np.max(x, axis=1, keepdims=True)
+    e_x = np.exp(x)
+    return e_x / e_x.sum(axis=1, keepdims=True)
+
+
+class ScaledDotProductAttention:
+    """
+    This is a simple class which demonstrates how we can implement Scaled Dot-Product Attention using numpy. It's not optimised at all.
+    """
+
+    def __init__(self, d_k: int, d_v: int, d_input: int) -> None:
+        """
+        We initialise our Q,K and V matrices
+
+        Note here that
+
+        d_k : Dimensionality of K and Q weight matrix
+        d_v : Dimensionality of V weight matrix
+        d_input : Dimensionality of embedding layer
+        """
+        self.w_q = np.random.randn(d_input, d_k)
+        self.w_k = np.random.randn(d_input, d_k)
+        self.w_v = np.random.randn(d_input, d_v)
+        self.d_k = d_k
+        self.d_input = d_input
+
+    def __call__(self, x: np.array, mask=False) -> np.array:
+        """
+        This is how we can call our forward pass function
+        """
+        assert (
+            x.shape[1] == self.d_input
+        ), f"Passed in an array of size {len(x)}. Maximum context size is {self.d_input}"
+        q_prime = x @ self.w_q.T  # x : (d_input , d_k ) , w_q.T : (d_k , d_input )
+        k_prime = x @ self.w_k.T  # x : (d_input , d_k ) , w_k.T : (d_k , d_input )
+        v_prime = x @ self.w_v.T
```

Comment:
little doubt: isn't this part of the multi-head attention? 🤔

(as of my understanding) scaled dot product attention just takes {Q, K, V} and operates on them as they are. From the paper I understood that the projection onto other spaces (so the multiplication with {W_Q, W_K, W_V}) happens before the self-attention is called 

![image](https://github.com/ivanleomk/llm-paper-club-asia-notes/assets/104844027/7562b99f-f48b-47f5-bc48-66c37082af4e)

Reply:
Hey @paulinamoskwa, on closer look, yes you're right, I confused Self-Attention and Scaled Dot Product Attention. 

Scaled dot product attention just takes {Q, K, V} and operates on them as they are. Self-Attention on the other hand takes in an input vector x with dim ( n, d_embeddings ) and then outputs a final output embedding of dim (n, d_embddings) with the help of the linear layer which does a transformation from the output of the Scaled Dot Product Attention 

Seems like my code here implements a Multi-Head attention block with a single step ( without the linear layer ) instead. I wanted to follow Raschka's implementation of Self-Attention from this [Video](https://www.youtube.com/watch?v=0PjHri8tc1c) which has this specific graphic

![CleanShot 2024-01-21 at 08 38 57](https://github.com/ivanleomk/llm-paper-club-asia-notes/assets/45760326/9215b898-ff76-4a50-9372-c79dc05be598)

Thanks for the feedback!